Example demonstrating Kanister integration with MySQL (#244)
kanisterio/kanister: examples/stable/mysql/README.md

# MySQL

[MySQL](https://MySQL.org) is one of the most popular database servers in the world. Notable users include Wikipedia, Facebook and Google.

## Introduction

This chart bootstraps a single node MySQL deployment on a [Kubernetes](http://kubernetes.io) cluster using the [Helm](https://helm.sh) package manager.

## Prerequisites

- Kubernetes 1.6+ with Beta APIs enabled
- PV provisioner support in the underlying infrastructure

> - Kanister version 0.21.0 installed

Reviewer: ```suggestion
- Kanister controller version 0.21.0 installed in your cluster
```

- Kanctl CLI installed (https://docs.kanister.io/tooling.html#kanctl)

## Installing the Chart

To install the chart with the release name `my-release`:

```bash
$ helm repo add stable https://kubernetes-charts.storage.googleapis.com/
$ helm repo update

$ helm install stable/mysql -n my-release --namespace mysql-test \
    --set mysqlRootPassword='asd#45@mysqlEXAMPLE' \
    --set persistence.size=10Gi
```

The command deploys an instance MySQL in `mysql-test` namespace on Kubernetes cluster with default configurations.

By default a random password will be generated for the root user. For setting your own password, use the `mysqlRootPassword` param as shown above.

You can retrieve your root password by running the following command. Make sure to replace [YOUR_RELEASE_NAME]:

>     printf $(printf '\%o' `kubectl get secret [YOUR_RELEASE_NAME]-mysql -o jsonpath="{.data.mysql-root-password[*]}"`)

Reviewer: Let's use base64 decode. 
```suggestion
   `kubectl get secret my-release-mysql -o jsonpath="{.data.mysql-root-password[*]}" | base64 --decode`
```

> **Tip**: List all releases using `helm list`

## Integrating with Kanister

If you have deployed MySQL application with name other than `my-release` and namespace other than `mysql-test`, you need to modify the commands(backup, restore and delete) used below to use the correct name and namespace

### Create Profile
Create Profile CR if not created already

```bash
$ kanctl create profile s3compliant --access-key 'AKIAIOSFODNN7EXAMPLE' \
	--secret-key 'wJalrXUtnFEMI%K7MDENG%bPxRfiCYEXAMPLEKEY' \

> 	--bucket <s3-bucket-name> --region ap-south-1 \

Reviewer: Let's be consistent with mixing string values and `<param>` style params

	--namespace mysql-test
```

**NOTE:**

The command will configure a location where artifacts resulting from Kanister
data operations such as backup should go. This is stored as a `profiles.cr.kanister.io`
*CustomResource (CR)* which is then referenced in Kanister ActionSets. Every ActionSet
requires a Profile reference to complete the action. This CR (`profiles.cr.kanister.io`)
can be shared between Kanister-enabled application instances.

### Create Blueprint

Create Blueprint in the same namespace as the controller

```bash
$ kubectl create -f kanister/mysql-blueprint.yaml -n kasten-io
```

Once MySQL is running, you can populate it with some data. Let's add a table called "pets" to a test database:

```bash
# Connect to MySQL by running a shell inside MySQL's pod
$ kubectl exec -ti $(kubectl get pods -n mysql-test --selector=app=my-release-mysql -o=jsonpath='{.items[0].metadata.name}') -n mysql-test -- bash

# From inside the shell, use the mysql CLI to insert some data into the test database
# Create "test" db

> $ mysql --user=root --password=asd#45@mysqlEXAMPLE

Reviewer: [minor]
```suggestion
$ mysql --user=root --password='asd#45@mysqlEXAMPLE'
```

```

Once the ActionSet status is set to "complete", you can see that the data has been successfully restored to MySQL

```bash
mysql> SHOW DATABASES;
+--------------------+
| Database           |
+--------------------+
| information_schema |
| mysql              |
| performance_schema |
| sys                |
| test               |
+--------------------+
5 rows in set (0.00 sec)

mysql> USE test;
Reading table information for completion of table and column names
You can turn off this feature to get a quicker startup with -A

Database changed
mysql> SHOW TABLES;
+----------------+
| Tables_in_test |
+----------------+
| pets           |
+----------------+
1 row in set (0.00 sec)

mysql> SELECT * FROM pets;
+----------+-------+---------+------+------------+-------+
| name     | owner | species | sex  | birth      | death |
+----------+-------+---------+------+------------+-------+
| Puffball | Diane | hamster | f    | 1999-03-30 | NULL  |
+----------+-------+---------+------+------------+-------+
1 row in set (0.00 sec)

```

## Troubleshooting

If you run into any issues with the above commands, you can check the logs of the controller using:

> ```bash

Reviewer: let's also describe the blueprint to get events

$ kubectl --namespace kasten-io logs -l app=kanister-operator
```

## Cleanup

### Uninstalling the Chart

To uninstall/delete the `my-release` deployment:

```bash

> $ helm delete my-release

Reviewer: ```suggestion
$ helm delete my-release --purge
```

# MySQL

[MySQL](https://MySQL.org) is one of the most popular database servers in the world. Notable users include Wikipedia, Facebook and Google.

## Introduction

This chart bootstraps a single node MySQL deployment on a [Kubernetes](http://kubernetes.io) cluster using the [Helm](https://helm.sh) package manager.

## Prerequisites

- Kubernetes 1.6+ with Beta APIs enabled
- PV provisioner support in the underlying infrastructure
- Kanister version 0.21.0 installed
- Kanctl CLI installed (https://docs.kanister.io/tooling.html#kanctl)

## Installing the Chart

To install the chart with the release name `my-release`:

```bash
$ helm repo add stable https://kubernetes-charts.storage.googleapis.com/
$ helm repo update

$ helm install stable/mysql -n my-release --namespace mysql-test \
    --set mysqlRootPassword='asd#45@mysqlEXAMPLE' \
    --set persistence.size=10Gi
```

> The command deploys MySQL on the Kubernetes cluster in the default configuration.

Reviewer 2: ```suggestion
The command deploys an instance MySQL in `mysql-test` namespace on Kubernetes cluster with default configurations.
```

> By default a random password will be generated for the root user. If you'd like to set your own password change the mysqlRootPassword

Reviewer 2: ```suggestion
By default a random password will be generated for the root user. For setting your own password, use the `mysqlRootPassword` param as shown above.
```

> in the values.yaml.

Reviewer 2: ```suggestion
```

You can retrieve your root password by running the following command. Make sure to replace [YOUR_RELEASE_NAME]:

    printf $(printf '\%o' `kubectl get secret [YOUR_RELEASE_NAME]-mysql -o jsonpath="{.data.mysql-root-password[*]}"`)

> **Tip**: List all releases using `helm list`

## Integrating with Kanister

> If you have deployed postgresql application with name other than `my-release` and namespace other than `mysql-test`, you need to modify the commands used below to use the correct name and namespace

Reviewer 2: ```suggestion
If you have deployed MySQL application with name other than `my-release` and namespace other than `mysql-test`, you need to modify the commands(backup, restore and delete) used below to use the correct name and namespace
```
